File: backend/python-ai-service/app/core/envelopes.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Generic, TypeVar
from uuid import uuid4

from pydantic import BaseModel, Field

T = TypeVar("T")


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


class ApiSuccess(BaseModel, Generic[T]):
    ok: bool = True
    data: T
    correlationId: str = Field(default_factory=lambda: str(uuid4()))
    timestamp: str = Field(default_factory=_now_iso)


class ApiErrorDetail(BaseModel):
    code: str
    message: str
    details: dict[str, Any] | None = None


class ApiError(BaseModel):
    ok: bool = False
    error: ApiErrorDetail
    correlationId: str = Field(default_factory=lambda: str(uuid4()))
    timestamp: str = Field(default_factory=_now_iso)


class BackendEvent(BaseModel, Generic[T]):
    id: str = Field(default_factory=lambda: f"evt_{uuid4()}")
    type: str
    version: str = "1.0"
    timestamp: str = Field(default_factory=_now_iso)
    correlationId: str = Field(default_factory=lambda: str(uuid4()))
    requestId: str | None = None
    sessionId: str | None = None
    source: str = "backend"
    payload: T


def success(data: Any, correlation_id: str | None = None) -> dict:
    return {
        "ok": True,
        "data": data,
        "correlationId": correlation_id or str(uuid4()),
        "timestamp": _now_iso(),
    }


def error(
    code: str,
    message: str,
    details: dict | None = None,
    correlation_id: str | None = None,
) -> dict:
    return {
        "ok": False,
        "error": {
            "code": code,
            "message": message,
            "details": details,
        },
        "correlationId": correlation_id or str(uuid4()),
        "timestamp": _now_iso(),
    }


def new_event(
    event_type: str,
    payload: Any,
    correlation_id: str | None = None,
    request_id: str | None = None,
    session_id: str | None = None,
) -> dict:
    return {
        "id": f"evt_{uuid4()}",
        "type": event_type,
        "version": "1.0",
        "timestamp": _now_iso(),
        "correlationId": correlation_id or str(uuid4()),
        "requestId": request_id,
        "sessionId": session_id,
        "source": "backend",
        "payload": payload,
    }

File: backend/go-gateway/internal/contracts/envelopes.go
package contracts

import (
	"encoding/json"
	"net/http"
	"time"
)

// ─── Canonical Response Envelopes ─────────────────────────────────────────────

type successEnvelope struct {
	Ok            bool            `json:"ok"`
	Data          any             `json:"data"`
	CorrelationID string          `json:"correlationId"`
	Timestamp     string          `json:"timestamp"`
}

type errorDetail struct {
	Code    string         `json:"code"`
	Message string         `json:"message"`
	Details map[string]any `json:"details,omitempty"`
}

type errorEnvelope struct {
	Ok            bool        `json:"ok"`
	Error         errorDetail `json:"error"`
	CorrelationID string      `json:"correlationId"`
	Timestamp     string      `json:"timestamp"`
}

// ─── BackendEvent — canonical SSE / WebSocket event ───────────────────────────

type BackendEvent struct {
	ID            string          `json:"id"`
	Type          string          `json:"type"`
	Version       string          `json:"version"`
	Timestamp     string          `json:"timestamp"`
	CorrelationID string          `json:"correlationId"`
	RequestID     *string         `json:"requestId"`
	SessionID     *string         `json:"sessionId"`
	Source        string          `json:"source"`
	Payload       json.RawMessage `json:"payload"`
}

// ─── Response helpers ─────────────────────────────────────────────────────────

func WriteSuccess(w http.ResponseWriter, correlationID string, data any) {
	env := successEnvelope{
		Ok:            true,
		Data:          data,
		CorrelationID: correlationID,
		Timestamp:     now(),
	}
	writeJSON(w, http.StatusOK, env)
}

func WriteError(w http.ResponseWriter, correlationID string, status int, code, message string, details map[string]any) {
	env := errorEnvelope{
		Ok: false,
		Error: errorDetail{
			Code:    code,
			Message: message,
			Details: details,
		},
		CorrelationID: correlationID,
		Timestamp:     now(),
	}
	writeJSON(w, status, env)
}

func WriteBadRequest(w http.ResponseWriter, correlationID, code, message string) {
	WriteError(w, correlationID, http.StatusBadRequest, code, message, nil)
}

func WriteNotFound(w http.ResponseWriter, correlationID string) {
	WriteError(w, correlationID, http.StatusNotFound, "not_found", "The requested resource was not found.", nil)
}

func WriteServiceUnavailable(w http.ResponseWriter, correlationID, service string) {
	WriteError(w, correlationID, http.StatusServiceUnavailable, "service_unavailable",
		service+" is not available", map[string]any{"service": service})
}

// WriteCircuitOpen writes a 503 with code "circuit_open" — use when the
// circuit breaker has tripped for a downstream service.
func WriteCircuitOpen(w http.ResponseWriter, correlationID, service string) {
	WriteError(w, correlationID, http.StatusServiceUnavailable, "circuit_open",
		service+" circuit breaker is open — too many recent failures, try again later",
		map[string]any{"service": service})
}

func WriteInternalError(w http.ResponseWriter, correlationID string) {
	WriteError(w, correlationID, http.StatusInternalServerError, "internal_error", "An internal error occurred.", nil)
}

func writeJSON(w http.ResponseWriter, status int, v any) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.Header().Set("X-Content-Type-Options", "nosniff")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(v)
}

func now() string {
	return time.Now().UTC().Format(time.RFC3339Nano)
}
